### ISSUE TITLE
C/C++/Rust implementation of Scaler's scheduler

### ISSUE THREAD

**Issue by rafa-be** [Citi/scaler#19]
A C, C++ or Rust implementation of Scaler's scheduler will have a few performance advantages, mainly by relying on a more efficient memory management, and by avoiding unnecessary copies of objects.

Contrary to the client and the worker implementation, re-implementing the scheduler in a language different from Python is less challenging, as it's not actively interacting with Python objects.

The scheduler is also the main scalability blocking piece, which makes its efficiency more critical in making Scaler more scalable.

**Comment by bcncpp** [comment]
Hello, i am also interested on this. If we go to the rust path we could use maturin, if we go to c++, pybind. Let's discuss here your plans and add a design.

**Comment by rafa-be** [comment]
Hi @bcncpp,

At the moment, our plans are to fully rewrite the scheduler in C++ or Rust, without any Python code or binding.

The current implementation already communicates with [0mq](https://zeromq.org/) and [Cap’n Proto](https://capnproto.org/) messages, which are not tied to any specific language.

However, we are still experimenting with new features to the Scaler's scheduler, so this is a mid-to-long term project.

Do you already have performance issues with the current Python implementation?

Raphael.

**Comment by bcncpp** [comment]
No, it is something that I can volunteer. it is a good opportunity to me. I provide a draft pr, so slowly i can iterate on your supervision.

**Comment by rafa-be** [comment]
> Hi [@rafa-be](https://github.com/rafa-be) No, it is something that I can volunteer. it is a good opportunity to me. I provide a draft pr, so slowly i can iterate on your supervision.

Please do 😃 . I'd suggest you keep the implementation as separated as possible from Python's (possibly even creating a new repository), limiting the dependency to the Can'p protocol files.

**Comment by bcncpp** [comment]
I use my fork, i create a folder native_scheduler that does the game,

**Comment by bcncpp** [comment]
![Image](https://github.com/user-attachments/assets/095c5a21-6798-4f29-b0b3-72cba8a28f3d)

**Comment by bcncpp** [comment]
1. `Scheduler` uses:
   - `SchedulerConfig` for configuration.
   - `ZMQConfig` for monitor address setup.
   - `AsyncBinder` to bind to the main address.
   - `AsyncConnector` for monitor connections.
   - `VanillaClientManager` to manage client interactions.
   - `VanillaObjectManager` for object-related operations.
   - `VanillaGraphTaskManager` for managing graph tasks.
   - `VanillaTaskManager` for task management.
   - `VanillaWorkerManager` to manage worker interactions.
   - `StatusReporter` to report statuses.

2. All manager classes (`Vanilla*Manager`) and `StatusReporter` are registered with dependencies like:
   - `binder`
   - `binder_monitor`
   - Other relevant managers (e.g., `client_manager`, `task_manager`).

**Comment by bcncpp** [comment]
Hi, I have an update. We've started to design something similar but in Rust. It's called streamgale, we've created an OS community around that. The initial sketch isL
# StreamGale Arch 1 - Architecture Diagram (Markdown Draw)

Virgil (SuperNode / Controller)
┌───────────────────────────────────────────────────────┐
│                    Virgil (Controller)                │
│  ┌──────────────────────────────────────────────────┐ │
│  │       Distributed Hash Table (DHT) Ring          │ │
│  │          (Consistent Hashing)                    │ │
│  └──────────────────────────────────────────────────┘ │
│                                                       │
│  ┌──────────────────────────────────────────────────┐ │
│  │     Registry Serve Node (Discovery, Coord.)      │ │
│  └──────────────────────────────────────────────────┘ │
│                                                       │
│  ┌──────────────────────────────────────────────────┐ │
│  │         TopologyGrpcServer (gRPC Comm)           │ │
│  └──────────────────────────────────────────────────┘ │
│                                                       │
│  ┌──────────────┐     ┌──────────────┐                │
│  │ SuperVisor 1 │ ... │ SuperVisor N │ ← Topology     │
│  └──────────────┘     └──────────────┘   Servers       │
└───────────────────────────────────────────────────────┘


Topology Server (SuperVisor)
┌───────────────────────────────────────────────────────┐
│                 Topology Server (SuperVisor)          │
│                                                       │
│  ┌──────────────────────────────────────────────────┐ │
│  │         Embedded HTTPS/S3 Topology Server        │ │
│  └──────────────────────────────────────────────────┘ │
│                                                       │
│  ┌────────────────────┐   ┌────────────────────────┐  │
│  │  CPU Worker Pool   │   │   GPU Worker Pool      │  │
│  └────────────────────┘   └────────────────────────┘  │
│           ▲                      ▲                    │
│           │                      │                    │
│    ┌─────────────────────────────────────────────┐    │
│    │       Dynamic Worker Allocator              │    │
│    └─────────────────────────────────────────────┘    │
│                                                       │
│    ┌─────────────────────────────────────────────┐    │
│    │      Docker-based Execution Engine          │    │
│    │   - Pulls Docker Images                     │    │
│    │   - Executes Commands in Containers         │    │
│    │   - Reports Telemetry via OpenTelemetry     │    │
│    └─────────────────────────────────────────────┘    │
└───────────────────────────────────────────────────────┘

We would like to hear from you about your use cases and we can help.
